feat: Add a process clock to Db and use it for Sequenced Entries

carols10cents opened the pull request:
This connects to #1157. I *think* it might finish it, except the ticket mentions the logical clock having the ability to be "updated to a new value (if passed value is > the current one)", but that ability wouldn't be used anywhere just yet, so I'd rather add it when it's needed. If I should do it now instead, let me know!

I tried having the `process_clock` field be an `AtomicU64`, but I don't think it works because there's not a way to read the value, compare it + 1 to another value, and possibly set the value with a guarantee that the value hasn't been modified between the read + set. Let me know if I'm missing some trick with atomics, this is a lower level than I usually work at 😬

tustvold commented:
Not looked too closely at the code, will leave that to someone who is more familiar with what is going on here - I thought we'd remove the system timestamp part :shrug: But thought I'd mention that is should be possible to do this with a loop that performs an atomic load and then a compare exchange, but not sure the codepath will be hot enough for this to actually matter...

alamb commented:
> read the value, compare it + 1 to another value, and possibly set the value with a guarantee that the value hasn't been modified between the read + set. 

I wonder if you can do that with `compare_exchange`? you get the previous value, calculate the new value, and then only update the atomic if it still has the previous value

https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.compare_exchange

alamb commented on server/src/db.rs:
```diff
@@ -465,6 +461,10 @@ impl Db {
         );
 
+        let process_clock = Arc::new(Mutex::new(
+            NonZeroU64::new(now_nanos()).expect("current time should not be 0"),
```
![image](https://user-images.githubusercontent.com/490673/117372566-b979e400-ae97-11eb-8295-89d35c0e86a0.png)

alamb commented on server/src/db.rs:
```diff
@@ -303,8 +299,8 @@ pub struct Db {
 
-    /// Used to allocated sequence numbers for writes
-    sequence: AtomicU64,
+    /// Process clock used in establishing a partial ordering of operations via a Lamport Clock.
```
```suggestion
    /// Process clock used in establishing a partial ordering of operations via a Lamport Clock. 
    /// Value is nanoseconds since the Unix Epoch
```

alamb commented on server/src/db.rs:
```diff
@@ -1369,8 +1388,8 @@ mod tests {
 
-    #[tokio::test]
-    async fn write_no_mutable_buffer() {
+    #[test]
```
👍  much nicer

alamb commented on server/src/db.rs:
```diff
@@ -2684,12 +2703,112 @@ mod tests {
+        let entry = lp_to_entry("cpu bar=1 10");
+        db.store_entry(entry).unwrap();
+
+        let between: u64 = Utc::now().timestamp_nanos().try_into().unwrap();
```
I wonder if you could use `now_nanos` defined above to get the current timestamp and reduce some of the replication in the tests (not that it really matters)

alamb reviewed (approved):
I think the PR and code implements what is intended. 

> but that ability wouldn't be used anywhere just yet, so I'd rather add it when it's needed.
Makes a lot of sense to me

e-dard reviewed (approved):
Nice 👍

domodwyer reviewed (approved):
Looks good! As the others said, a CAS loop might be a nice addition but certainly good as is 👍 

The devil is most certainly in the details - making sure we inc. this counter in response to incoming replicated writes from other nodes, max()-ing it and propagating the value to others will be an important implementation too!

carols10cents commented:
> I thought we'd removed the system timestamp part 🤷

@tustvold This is from the design in the "Replicating and Sequencing Writes" google doc that starts with a date of March 22nd... I'm not looking at an outdated design doc, am I...?

shepmaster commented on server/src/db.rs:
```diff
@@ -1055,14 +1053,30 @@ impl Db {
+            result = self.try_update_process_clock();
+        }
+
+        let next = result.unwrap();
```
I think this can be written as

```rust
    let next = loop {
        if let Ok(next) = self.try_update_process_clock() {
            break next;
        }
    };
```

Would help with repeating the function call and the `unwrap`

alamb commented on server/src/db.rs:
```diff
@@ -1055,14 +1053,30 @@ impl Db {
-        ClockValue::new(next)
+
+        self.process_clock
+            .compare_exchange(
```
I reviewed this logic and it makes sense to me

mkmik reviewed (approved):
LGTM!

pauldix reviewed (approved):
LGTM. Is the process clock something that would be a little cleaner as its own new type rather than the functions off DB around the AtomicU64?

crepererum commented on server/src/db.rs:
```diff
@@ -1048,9 +1048,35 @@ impl Db {
+    fn try_update_process_clock(&self) -> Result<u64, u64> {
+        let now = now_nanos();
+        let current_process_clock = self.process_clock.load(Ordering::SeqCst);
+        let next_candidate = current_process_clock + 1;
```
Is the idea here that the clock must have advanced by at least 1ns after the last update? Let's check (I might be wrong though):

- If you have a 1GHz processor
- you have 1G cycles per second
- makes 1 cycle every 1/1G s
- makes 1 cycle every 1/1M millisecond
- makes 1 cycle every 1/1K microsecond
- makes 1 cycle every nanosecond

With super-scalar processors (i.e. having multiple instructions in parallel on the same core) and higher clock speeds, you probably still don't get below 1ns per call (esp. with `now_nanos` and the compare-exchange call). However if you have multiple cores at the same time updating the process clock, you basically take this 1ns-bet per core and with a server-processor of 128 cores that gives you 128ns per call which then is already in range of "could be wrong".

So my point is: why adding this 1ns in the first place?

crepererum replied:
Oh, is the point that the clock should always increase and never stay the same?

mkmik replied:
It's very unlikely but not impossible that the system clock gets reset it slewed in just the right way for it to return to exactly the same value as it was when we checked it last time.

System clocks suck. But we decided to use to superimpose the system clock value domain on top of our Lamport clock in a way that is technically sound as far as the Lamport clock goes, and yet it lets us roughly tell the time a given event occurred when we state at the logical clock values.

That is, when all things are going well. For example all it takes is one of the replicas to accidentally misconfigure their clock to say 2060 and even after it gets fixed, and goes back to present, the Lamport clock must be monotonic hence it must increment by one unit (not really a nanosecond) per event.

crepererum replied:
@mkmik thanks for the insightful explanation :)

crepererum commented on server/src/db.rs:
```diff
@@ -2684,12 +2719,110 @@ mod tests {
+
+        db.process_clock.store(later, Ordering::SeqCst);
+
+        // Every call to next_process_clock should increment at least 1, even in this case
```
Even though it would not catch all errors, we could at least have a multi-thread test (e.g. with 10 threads, each trying to update the clock 10 times checking the clock increments).

crepererum commented on server/src/db.rs:
```diff
@@ -1307,6 +1333,15 @@ impl CatalogProvider for Db {
+// Convenience function for getting the current time in a `u64` represented as nanoseconds since
+// the epoch
+fn now_nanos() -> u64 {
+    Utc::now()
```
BTW: this clock might jump backwards if for example the system picks up a new time via NTP or if the hardware / supervisor (VM) clock is modified. With the above logic that should be totally fine, however I would like to see at least a comment acknowledging it, esp. since this is in contrast to the monotonic clock that Rust would give us (which however wouldn't be in UTC).

mkmik replied:
Yes. We have a design doc covering all this stuff and hence it's easy now to fall in the trap of taking things for granted. The code comments should repeat those assumptions or even  better the function and type names should make that explicit, e.g. system_clock_now

carols10cents commented:
@pauldix @mkmik @crepererum I extracted the process clock to its own type, database is getting rather large :) I also hopefully captured some of the assumptions of the design in the name and comment improvements you all suggested, and added a test where multiple threads are accessing one process clock and asserting it always increases.

Would any of you mind taking another look at the new commits? Thank you for the great suggestions!
